src/ripple/shamap/impl/SHAMap.cpp

```diff
@@ -419,28 +423,25 @@ SHAMap::unshareNode (std::shared_ptr<Node> node, SHAMapNodeID const& nodeID)
 SHAMapTreeNode*
 SHAMap::firstBelow(SHAMapAbstractNode* node, NodeStack& stack) const
 {
-    // Return the first item below this node
-    while (true)
+    // Return the first item at or below this node
+    assert(node != nullptr);
+    if (node->isLeaf())
+        return static_cast<SHAMapTreeNode*>(node);
+    auto inner = static_cast<SHAMapInnerNode*>(node);
+    for (int i = 0; i < 16; ++i)
     {
-        assert(node != nullptr);
-        if (node->isLeaf())
-            return static_cast<SHAMapTreeNode*>(node);
-        bool foundNode = false;
-        auto inner = static_cast<SHAMapInnerNode*>(node);
-        for (int i = 0; i < 16; ++i)
+        if (!inner->isEmptyBranch(i))
         {
-            if (!inner->isEmptyBranch(i))
-            {
-                node = descendThrow(inner, i);
-                assert(!stack.empty());
-                stack.push({node, stack.top().second.getChildNodeID(i)});
-                foundNode = true;
-                break;
-            }
+            node = descendThrow(inner, i);
+            assert(!stack.empty());
+            stack.push({node, stack.top().second.getChildNodeID(i)});
+            if (node->isLeaf())
+                return static_cast<SHAMapTreeNode*>(node);
+            inner = static_cast<SHAMapInnerNode*>(node);
+            i = -1;  // scan all 16 branches of this new node
```

Comment:
I originally looked at it with a bit of surprise.  But the previous version had an outer `while (true)` which this removes.  I also thought of changing the `for` loop to a `while` which would be more idiomatic.  But then `i` would need to be lofted outside of the loop.

For me the comment on this line was sufficient.  But Mr. D's suggestion of a note at the top of the loop might be a good thing for a casual reader.